File: app/src/main/java/br/com/douglas/fukuhara/lodjinha/adapter/ProductsListAdapter.java
package br.com.douglas.fukuhara.lodjinha.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import java.math.RoundingMode;
import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;

import br.com.douglas.fukuhara.lodjinha.R;
import br.com.douglas.fukuhara.lodjinha.network.vo.ProductDataVo;

public class ProductsListAdapter extends RecyclerView.Adapter<ProductsListAdapter.HomeBestSellerViewHolder> {

    private List<ProductDataVo> mBestSellerProductsList;
    private final HomeBestSellerClickListener mListener;

    public ProductsListAdapter(List<ProductDataVo> bestSellerProductsList, HomeBestSellerClickListener listener) {
        mBestSellerProductsList = bestSellerProductsList;
        mListener = listener;
    }

    @NonNull
    @Override
    public HomeBestSellerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        return new HomeBestSellerViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.best_seller_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull HomeBestSellerViewHolder holder, int position) {
        if (getItemCount() > 0) {
            holder.onBind(mBestSellerProductsList.get(position));
        }
    }

    @Override
    public int getItemCount() {
        return mBestSellerProductsList == null ? 0 : mBestSellerProductsList.size();
    }

    public void updateDataSetContent(List<ProductDataVo> list) {
        mBestSellerProductsList = list;
        notifyDataSetChanged();
    }

    public interface HomeBestSellerClickListener {
        void onBestSellerItemClick(ProductDataVo productDataVo);
    }

    class HomeBestSellerViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final ViewGroup mViewGroup;
        private final ImageView ivProductImage;
        private final TextView tvProductName;
        private final TextView tvPrevPrice;
        private final TextView tvFinalPrice;
        private final NumberFormat mNumberFormatter;

        HomeBestSellerViewHolder(@NonNull View itemView) {
            super(itemView);

            mViewGroup = (ViewGroup) itemView;
            ivProductImage = itemView.findViewById(R.id.iv_best_seller_list_image);
            tvProductName = itemView.findViewById(R.id.tv_best_seller_list_prod_name);
            tvPrevPrice = itemView.findViewById(R.id.tv_best_seller_list_prev_price);
            tvFinalPrice = itemView.findViewById(R.id.tv_best_seller_list_final_price);

            mNumberFormatter = NumberFormat.getInstance(new Locale("pt", "BR"));
            mNumberFormatter.setRoundingMode(RoundingMode.HALF_UP);
            mNumberFormatter.setMaximumFractionDigits(2);
            mNumberFormatter.setMinimumFractionDigits(2);

            itemView.setOnClickListener(this);
        }

        void onBind(ProductDataVo productDataVo) {
            String urlImagem = productDataVo.getUrlImagem();
            String productName = productDataVo.getNome();

            // TODO: Configure: fallback, onError, placeholder
            Glide
                    .with(mViewGroup)
                    .load(urlImagem)
                    .placeholder(R.mipmap.image_placeholder)
                    .error(R.mipmap.image_placeholder_error)
                    .into(ivProductImage);

            tvProductName.setText(productName);
            String productPrecoDe = mNumberFormatter.format(productDataVo.getPrecoDe());
            String productPrecoPor = mNumberFormatter.format(productDataVo.getPrecoPor());
            tvPrevPrice.setText(mViewGroup.getContext().getString(R.string.best_seller_prev_price, productPrecoDe));
            tvFinalPrice.setText(mViewGroup.getContext().getString(R.string.best_seller_final_price, productPrecoPor));
        }

        @Override
        public void onClick(View view) {
            int clickedPosition = getAdapterPosition();
            mListener.onBestSellerItemClick(mBestSellerProductsList.get(clickedPosition));
        }
    }
}

File: Lodjinha/app/src/main/kotlin/com/b2w/lodjinha/ui/productslist/ProductsListAdapter.kt
package com.b2w.lodjinha.ui.productslist

import android.graphics.Paint
import android.support.constraint.ConstraintLayout
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.b2w.lodjinha.R
import com.b2w.lodjinha.data.model.ProductItem
import com.b2w.lodjinha.util.toCurrency
import com.squareup.picasso.Picasso

class ProductsListAdapter(
    var items: MutableList<ProductItem>,
    private val onItemClickListener: (items: MutableList<ProductItem>, position: Int, view: View) -> Unit
) : RecyclerView.Adapter<ProductsListViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductsListViewHolder {
        return ProductsListViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.view_product_item,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: ProductsListViewHolder, position: Int) {
        holder.bind(items[position]) { view ->
            onItemClickListener(items, position, view)
        }
    }

    override fun getItemCount(): Int = items.size

    override fun getItemViewType(position: Int) = position

    fun addItems(newItems: List<ProductItem>) {
        items.addAll(newItems)
    }
}

open class ProductsListViewHolder(view: View) : RecyclerView.ViewHolder(view) {

    val container = view.findViewById<ConstraintLayout>(R.id.view_product_container)
    val imageview = view.findViewById<ImageView>(R.id.view_product_image)
    val description = view.findViewById<TextView>(R.id.view_product_description)
    val fromPrice = view.findViewById<TextView>(R.id.view_product_from_price)
    val toPrice = view.findViewById<TextView>(R.id.view_product_to_price)

    open fun bind(product: ProductItem, listener: (View) -> Unit) {
        description.text = product.description

        Picasso.get()
            .load(product.image)
            .placeholder(R.drawable.greyish_shape_oval)
            .error(R.drawable.greyish_shape_oval)
            .into(imageview)

        fromPrice.text = fromPrice.resources.getString(R.string.product_from_price, product.fromPrice.toCurrency())
        fromPrice.paintFlags = fromPrice.getPaintFlags() or Paint.STRIKE_THRU_TEXT_FLAG

        toPrice.text = fromPrice.resources.getString(R.string.product_to_price, product.toPrice.toCurrency())

        container.setOnClickListener { listener(container) }
    }
}